Update general UPP for GTG4

YaliMao-NOAA opened the pull request:
1. Change gtg_algo() interface in MDLFLD.f
2. Add CIT(:,:,:) array for In-Cloud Turbulence, ALLOCATE_ALL.f VRBLS3D_mod.f DEALLOCATE.f
3. Modify CMakeLists.txt for updated/new GTG source code files.
4. Add stub code for the new GTG source	files in general UPP
5. Modify interpolation	in MISCLN.f for	RRFS GTG
6. Modify jobs/J_NCEPPOST to run WAFS downstream only
   Add scripts/exwafs_atmos_nceppost.sh
   Add ush/wafs_nceppost.sh
7. Add GEFS WAFS control files (.xml .txt)
8. (As an update later on) Add gtg_on option
9. Remove WAFS fields on ICAO standard atmospheric levels from post_avblflds.xml
10. Remove MDL2STD_P.f which does interpolation on ICAO standard atmospheric levels which is not ICAO required.

Commit:
Add GEFS WAFS control file

Commit:
Merge remote-tracking branch 'upstream/develop' into v17

Commit:
Merge remote-tracking branch 'upstream/develop' into v17

Commit:
1. Change gtg_algo() interface in MDLFLD.f
2. Add CIT(:,:,:) array for In-Cloud Turbulence, ALLOCATE_ALL.f VRBLS3D_mod.f DEALLOCATE.f
3. Modify CMakeLists.txt for updated/new GTG source code files.
4. Add stub code for the new GTG source files in general UPP
5. Modify interpolation in MISCLN.f for RRFS GTG
6. Modify jobs/J_NCEPPOST to run WAFS downstream only
   Add scripts/exwafs_atmos_nceppost.sh
   Add ush/wafs_nceppost.sh

WenMeng-NOAA commented:
@YaliMao-NOAA Do you have an UPP issue linked this PR? If not, please open a new one.

YaliMao-NOAA commented:
@WenMeng-NOAA I opened an issue https://github.com/NOAA-EMC/UPP/issues/763

WenMeng-NOAA commented:
@YaliMao-NOAA Please add a change log with your name in the fortran code files where if change log sessions are  available.

YaliMao-NOAA commented:
> @YaliMao-NOAA Please add a change log with your name in the fortran code files where if change log sessions are available.

You mean the code change history?

WenMeng-NOAA commented:
@YaliMao-NOAA See the example at
https://github.com/NOAA-EMC/UPP/blob/develop/sorc/ncep_post.fd/INITPOST_NETCDF.f#L42

Commit:
1. Add updates to the log/history of the related Fortran source code
2. Add gtg_on to POSTGPVARS in scripts/exwafs_atmos_nceppost.sh,
   and add gtg_on to namelist 'nampgb' in WRFPOST.f

YaliMao-NOAA commented:
@WenMeng-NOAA I added change logs to the Fortran source code files. 
Also I added gtg_on option to the namelist

WenMeng-NOAA commented:
@YaliMao-NOAA My UPP with gtg code build process failed as:
```
CMake Error at sorc/ncep_post.fd/CMakeLists.txt:194 (add_library):
  Cannot find source file:

    post_gtg.fd/gtg_cdf_mod.F90

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .mpp .m .M .mm .h .hh .h++
  .hm .hpp .hxx .in .txx .f .F .for .f77 .f90 .f95 .f03 .ispc


CMake Error at sorc/ncep_post.fd/CMakeLists.txt:194 (add_library):
  No SOURCES given to target: upp


CMake Generate step failed.  Build files cannot be regenerated correctly
```
There is no file "gtg_cdf_mod.F90"  from gtg code checkout with the revision d0ad324.

YaliMao-NOAA commented:
Sorry, I should have update UPP_GTG develop branch. I am requesting a viewer to approve a merge pull request of GTG code.

Commit:
update gtg code revision

WenMeng-NOAA commented on sorc/ncep_post.fd/WRFPOST.f:
```diff
@@ -36,6 +36,8 @@
 !> 2023-04-04 | Li(Kate Zhang)  |Add namelist optoin for CCPP-Chem (UFS-Chem) 
 !         and 2D diag. output (d2d_chem) for GEFS-Aerosols and CCPP-Chem model.
 !> 2023-05-20 | Rahul Mahajan             | Bug fix for fileNameFlat as namelist configurable
+!> 2023-05-20 | Rahul Mahajan             | Bug fix for fileNameFlat as namelist configurable
```
@YaliMao-NOAA Please remove this duplicate line.

Commit:
Remove one duplicate line of comment

WenMeng-NOAA commented:
@YaliMao-NOAA Please verify my UPP standalone test results at /u/wen.meng/ptmp/gfs.2019083000 on Cactus and let me know if you saw issues.

WenMeng-NOAA commented:
@YaliMao-NOAA The UPP RT failed in all test case as:
```
+ mpiexec -l -n 96 -ppn 48 /u/wen.meng/ptmp/yali/UPP/exec/upp.x
nid003523.cactus.wcoss2.ncep.noaa.gov 2: forrtl: severe (174): SIGSEGV, segmentation fault occurred
Image              PC                Routine            Line        Source          
upp.x              0000000000F4F36A  Unknown               Unknown  Unknown
libpthread-2.31.s  0000150A256408C0  Unknown               Unknown  Unknown
upp.x              0000000000861FAA  mdl2p_                    373  MDL2P.f
libiomp5.so        0000150A26A6D3F3  __kmp_invoke_micr     Unknown  Unknown
libiomp5.so        0000150A269F1937  __kmp_fork_call       Unknown  Unknown
libiomp5.so        0000150A269B5533  __kmpc_fork_call      Unknown  Unknown
upp.x              000000000081986E  mdl2p_                    322  MDL2P.f

```
I guess you need to add new option "gtg_on"  in these kind of routine for skipping gtg related computations.

YaliMao-NOAA commented:
The test result looks good to me /u/wen.meng/ptmp/gfs.2019083000

As to the failure of SIGSEGV on your RT, you are right, gtg_on needs to be turned on by 'gtg_on=.true.' Do you think I should include 'gtg_on' option to avoid this kind of errors in mdl2p?

WenMeng-NOAA commented:
> The test result looks good to me /u/wen.meng/ptmp/gfs.2019083000
> 
> As to the failure of SIGSEGV on your RT, you are right, gtg_on needs to be turned on by 'gtg_on=.true.' Do you think I should include 'gtg_on' option to avoid this kind of errors in mdl2p?

@YaliMao-NOAA Yes, you might also check other gtg related routines.

Commit:
Fix bug of GTG interpolations in MDL2P.f
Remove WAFS fields on ICAO standard atmospheric levels from post_avblflds.xml
And remove MDL2STD_P.f which does interpolation on ICAO standard atmospheric levels

YaliMao-NOAA commented:
I modified the code, and added a new issue related to this PR. https://github.com/NOAA-EMC/UPP/issues/764

WenMeng-NOAA commented:
> I modified the code, and added a new issue related to this PR. #764

@YaliMao-NOAA Can you note your new changes at the beginning session of this PR, such as
1. Change gtg_algo() interface in MDLFLD.f
2. Add CIT(:,:,:) array for In-Cloud Turbulence, ALLOCATE_ALL.f VRBLS3D_mod.f DEALLOCATE.f
...

WenMeng-NOAA commented:
@YaliMao-NOAA You need to tune your branch further. Here are the latest errors from the UPP RT for nmmb:
```
upp.x              00000000008626A8  mdl2p_                    387  MDL2P.f
libiomp5.so        000014EA165423F3  __kmp_invoke_micr     Unknown  Unknown
libiomp5.so        000014EA164C6937  __kmp_fork_call       Unknown  Unknown
libiomp5.so        000014EA1648A533  __kmpc_fork_call      Unknown  Unknown
upp.x              0000000000819CBC  mdl2p_                    335  MDL2P.f
upp.x              0000000000612D37  process_                   78  PROCESS.f
upp.x              000000000056D7FC  MAIN__                    701  WRFPOST.f
```
The whole runtime log file can be found /u/wen.meng/noscrub/ncep_post/post_regression_test_new/jobs/out.nmmb.test.
You might run the UPP RT by yourself for debugging. The UPP RT procedure can be found at 
https://docs.google.com/document/d/1gtk1DBxEa9GsaWbxrxtB4Ot7U_lNIHjLXaU2F6FyD7M/edit

Commit:
Fix a bug by initializing "gtg_interpolation = .false." in MDL2P.f

YaliMao-NOAA commented:
@WenMeng-NOAA It turns out gtg_interpolation was not initialized in MDL2P.f, I fixed it. Thank you for recommendation to do the regression test by myself. All regression tests passed. FYI, for rtma test, should it be 'cd $workdir' instead of '$homedir/jobs'?

WenMeng-NOAA commented on sorc/ncep_post.fd/PROCESS.f:
```diff
@@ -115,11 +116,11 @@ SUBROUTINE PROCESS(kth,kpv,th,pv,iostatusD3D)
-      if(me==0) write(*,*) "PROCESS MDL2STD_P done"
-      MDL2STD_tim = MDL2STD_tim +(mpi_wtime() - btim)
-!
+!      btim = mpi_wtime()
```
@YaliMao-NOAA Could you remove these comment lines since the routine MDL2STD_P was deleted from UPP code in your PR?

Commit:
Remove comments and a variable related to MDL2STD_P
Piggyback changes of parm/makefile for cleaning out GEFS anal control file generation

WenMeng-NOAA commented:
@YaliMao-NOAA With new version of gtg4 implemented, do you think the gtg files gtg.config.gfs  gtg_imprintings.txt under UPP/parm should be removed?

Commit:
Remove GTG related parm files. Instead, scripts will copy the parm file from
sorc/ncep_post.fd/post_gtg.fd

WenMeng-NOAA commented:
The UPP RT is completed on WCOSS2 with no changed results.

WenMeng-NOAA commented:
@FernandoAndrade-NOAA You may start the UPP RT on Hera and Orion. There will no baseline update with this PR.

FernandoAndrade-NOAA commented:
Apologies for the delay, Orion UPP RTs have completed successfully with no changes to results. Hera should finish momentarily.

FernandoAndrade-NOAA commented:
Hera UPP RTs have completed as well with no changes to results.

WenMeng-NOAA commented:
@gspetro-NOAA Please be aware of the updates in parm/post_avblfld.xml in this PR for your new release of the UPP user guide later.

WenMeng-NOAA commented:
This PR is ready for merging.
